### PR TITLE
[Evals API] [1/n] Initial API

### PR DESCRIPTION
**Pull request by yanxi0830**
# TL;DR
* Initial Evals API

# High-Level Design

1. Matching registry API (datasets, scoring_functions) with action API (datasetio, scoring)

##### datasets <> datasetio
- datasets is a registry/routing_table managing all datasets (defined by DatasetDefWithProvider)
  - each DatasetDefWithProvider defines a schema the underlying dataset needs to obey
  - we have custom type with validators to validate correct input scorer columns
- datasetio is backed by concrete providers to expose get_rows_paginated() API for iterating over dataset rows. 

##### scoring_functions <> scoring
- scoring_functions is a registry/routing_table managing all scorers
- scoring is backed by concrete providers to get metrics

2. Evals API (/evals) for full benchmarking evals including model generation
- evals API takes in (dataset_id, scoring_functions, model) and run generation + scoring to output metrics. it can be async submitted as a job

# How it will be used

1. Benchmark evals: Standard benchmark with no custom model specific pre/post-processing for apple-to-apple comparison on models with different capabilities. Use identical inputs to models. 
  - use /eval to run full generation->scoring
2. Application evals:  End-to-end tests are executing the same application code with precreated inputs and outputs. 
 - use /scoring with formatted the input datasets generated from application

### CODE REVIEW

File: llama_stack/apis/datasetio/datasetio.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class DatasetStore(Protocol):
+    def get_dataset(self, identifier: str) -> DatasetDefWithProvider: ...
+
+
+@runtime_checkable
+class DatasetIO(Protocol):
+    # keeping for aligning with inference/safety, but this is not used
+    dataset_store: DatasetStore
```
Comment (yanxi0830) [inline llama_stack/apis/datasetio/datasetio.py:30]: Might not be used.

File: llama_stack/apis/datasets/datasets.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+class DatasetSchema(BaseModel):
+    columns: Dict[str, ParamType]
+
+
+@json_schema_type
+class DatasetDef(BaseModel):
+    identifier: str = Field(
+        description="A unique name for the dataset",
+    )
+    dataset_schema: DatasetSchema = Field(
```
Comment (ashwinb) [inline llama_stack/apis/datasets/datasets.py:28]: now wondering if we should just inline the `columns` property of schema into the `Def` itself?

File: llama_stack/apis/eval/eval.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    type: Literal["agent"] = "agent"
+    config: AgentConfig
+
+
+EvalCandidate = Annotated[
+    Union[ModelCandidate, AgentCandidate], Field(discriminator="type")
+]
+
+
+@json_schema_type
```
Comment (ashwinb) [inline llama_stack/apis/eval/eval.py:35]: can you move this to `apis/common/job_types.py`

**Review by ashwinb — APPROVED** [review]
